### ISSUE TITLE
Database Query Assembly Fault

### ISSUE THREAD

**Issue by bhsmither** [cubecart/v6#1698]
Example:
```
$sales = $GLOBALS['db']->select('CubeCart_order_summary', array('order_date', 'total'), array('order_date' => '>='.$last_year_start, 'status' => array(2, 3), 'total' => '>0'));
```
where `$last_year_start` is not defined.

The resulting query assembled is:
```
SELECT `order_date`, `total` FROM `CubeCart_order_summary` WHERE CubeCart_order_summary.order_date > '=' AND `status` IN (2,3) AND CubeCart_order_summary.total > '0' ;
```
The database engine will throw error 1292:
Truncated incorrect DOUBLE value: '='

The problem is in `Database->where()` (two places):
```
if (preg_match('#^([<>!~\+\-]=?)(.+)#', $key, $match)) {  // The $key starts with a special character with an optional =.
```
This works: `array('col' => '>=a')`
This does not: `array('col' => '>=')`
(The second instance of `preg_match()` uses $value.)

What happens is that the equal sign ends up in `$match[2]` and is `sqlSafe()`d.

As explained here (http://us1.php.net/manual/en/regexp.reference.repetition.php), the question mark quantifier causes the repetition parser to become _lazy_, while it is implied that the quantifier {0,1} keeps the parser _greedy_. (I have not yet verified this.)

Still, if this is fixed so that the result is:
```
WHERE CubeCart_order_summary.order_date >= _what???_

```
there is no follow-on test to make sure there is a valid argument in the expression held in `$match[2]`.

**Comment by bhsmither** [comment]
I think the changes in dashboard.index.inc.php fixes #1697, but does not fix this issue with where().

**Comment by abrookbanks** [comment]
It does I think. Can you show me if I'm wrong?

**Comment by bhsmither** [comment]
You have fixed sending bad data in the location in the code specified in #1697 .

That doesn't fix any other location where bad data may get sent.

That said, bad data is bad and where that happens needs to get fixed. But, the other side of the programming lesson-learned is to efficiently handle possible bad data received. (That's the second point I made.)

This issue references the database class, and should be evaluated on its own merits.

A query such as:
```
SELECT * WHERE `path` != '';    // not equal to a zero-length string - MySQL does not have IS [NOT] EMPTY
```
The call to `Database->select()` with the $where argument could be (and was the only way I knew at the time):
```
array('path' => "!=''")
```
This didn't work because the result is looking for actual apostrophe characters. I then tried:
```
array('path' => "!=")
```
This didn't work because the _laziness_ of the question mark put the equal sign in the second capture due to there being nothing else to use. *This preg_match() should have failed. It didn't.*

I finally found something that does work:
```
array('path' => "!= ")
```
This puts the space in the second capture, which gets trimmed and $value ends up being a zero-length-string.
